### ISSUE TITLE
Ocationally a bucket created locally with resen doesn't have "sudo" enabled. 

### ISSUE THREAD

**Issue by pmreyes2** [EarthCubeInGeo/resen#60]
Sometimes, I don’t know if randomly, or the first time after a reboot , when a bucket is created with resen, Jovyan sudo is not enabled.

**Comment by pmreyes2** [comment]
I can't reproduce the error now. I have created buckets with and without mounting storage, and sudo is available on both. Will come back here when I can reproduce the error.

**Comment by ljlamarche** [comment]
I've been able to reproduce the error at least twice.  I've noticed a similar issue.  I can use sudo in a bucket I created with resen several months ago, but not one I tried to create today.  This is what happens when I try to use sudo in the terminal of a newly-created bucket:

```
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

(py36) jovyan@b5809f2cf319:~$ sudo ls
[sudo] password for jovyan: 
jovyan is not in the sudoers file.  This incident will be reported.
```

**Comment by valentic** [comment]
What groups are you in? Run the "groups" command.

**Comment by ljlamarche** [comment]
I (user jovyan) am in groups users and sudo.

**Comment by valentic** [comment]
The next thing to look at then is the sudo config (/etc/sudoers) for the 
line starting %sudo

I don't have the new core running on the server yet, so I can't check it 
yet. The 2019 version didn't have jovyan in the sudo group.

Todd

**Comment by ljlamarche** [comment]
Here's the content of `/etc/sudoers` file (from the older bucket I CAN use sudo in):
```
#
# This file MUST be edited with the 'visudo' command as root.
#
# Please consider adding local content in /etc/sudoers.d/ instead of
# directly modifying this file.
#
# See the man page for details on how to write a sudoers file.
#
Defaults        env_reset
Defaults        mail_badpass
Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin"

# Host alias specification

# User alias specification

# Cmnd alias specification

# User privilege specification
root    ALL=(ALL:ALL) ALL

# Members of the admin group may gain root privileges
#%admin ALL=(ALL) ALL

# Allow members of group sudo to execute any command
%sudo   ALL=(ALL:ALL) ALL

# See sudoers(5) for more information on "#include" directives:

#includedir /etc/sudoers.d
```

**Comment by ljlamarche** [comment]
Also, I believe user jovyan is intentionally not in the sudo group on the server.  We didn't want users on the server to have root capabilities.

**Comment by pmreyes2** [comment]
@ljlamarche have you been able to create a new bucket locally that has no `sudo` enabled? For some reason my macbook freeze a couple of minutes ago and gave me an opportunity to try this again. So, the first bucket I created didn't have `sudo` enabled for jovyan user. Subsequently I created another bucket with exactly the same mount and this time `sudo` was enabled. And I can not reproduce the error again. Steps followed to reproduce error:

- Restart macbook
```
[resen] >>> create
Please enter a name for your bucket.
Valid names may not contain spaces and must start with a letter and be less than 20 characters long.
>>> Enter bucket name: bucket1
Please choose a version of resen-core.
Available versions: 2019.1.0
>>> Select a version: 2019.1.0
Local directories can be mounted to /home/jovyan/mount in a bucket.  You can specify either r or rw privileges for each directory mounted.  
>>> Mount storage to /home/jovyan/mount? (y/n): y
>>> Enter local path: /Users/LOCALUSERNAME/DATA
>>> Enter bucket path: /home/jovyan/mount/DATA
>>> Enter permissions (r/rw): rw
>>> Mount additional storage to /home/jovyan/mount? (y/n): n
>>> Start bucket and jupyterlab? (y/n): y
...adding core...
...adding ports...
...adding mounts...
Bucket created successfully!
...starting jupyterlab...
Jupyter lab can be accessed in a browser at: http://localhost:9008/?token=........
```

**Comment by asreimer** [comment]
The bug produces this error inside the docker container:

    ttyname failed: Inappropriate ioctl for device

The `sed -i` command needs a pseudo-TTY allocated, but by default the `exec_run` method has `tty=False`. See: https://docker-py.readthedocs.io/en/stable/containers.html?highlight=exec_run#docker.models.containers.Container.exec_run

Setting `tty=True` fixes the problem.

I troubleshooted this in python:

    import resen
    r = Resen()
    bucket = r.get_bucket('demo')
    container = r.dockerhelper.docker.containers.get(bucket['container'])
    cmd = "bash -cl 'sed -i \"s/^#\s*\(%sudo\s\+ALL=(ALL:ALL)\s\+ALL\)/\\1/\" /etc/sudoers '"
    result = container.exec_run(cmd,user='root',detach=False)
    result                                                                                                 
    Out[35]: ExecResult(exit_code=0, output=b'mesg: ttyname failed: Inappropriate ioctl for device\n')
    result = container.exec_run(cmd,user='root',detach=False,tty=True)
    result                                                                                                 
    Out[37]: ExecResult(exit_code=0, output=b'')


Bug fixed in commit: https://github.com/EarthCubeInGeo/resen/commit/09a520bac2429f0f02e9ade978ec0d629b9a7f77
